dom/src/EventDelegator.ts

```diff
@@ -53,116 +82,345 @@ function indexOf(arr: Array<Destination>, searchId: number): number {
+      this.nonBubblingListeners.set(
+        eventType,
+        new Map<Element, NonBubblingListener>(),
+      );
+    }
+    const map = this.nonBubblingListeners.get(eventType) as Map<
+      Element,
+      NonBubblingListener
+    >;
+    map.set(element, {sub, destination});
+  }
+
+  private resetEventListeners(): void {
+    const iter = this.domListeners.entries();
+    let curr = iter.next();
+    while (!curr.done) {
+      const [type, {sub, passive}] = curr.value;
+      sub.unsubscribe();
+      this.setupDOMListener(type, passive);
+      curr = iter.next();
+    }
+  }
+
+  private resetNonBubblingListeners(): void {
+    const newMap = new Map<string, Map<Element, NonBubblingListener>>();
+    const insert = makeInsert(newMap);
+
+    this.nonBubblingListeners.forEach((map, type) => {
+      map.forEach((value, elm) => {
+        if (!document.body.contains(elm)) {
+          const {sub, destination} = value;
+          if (sub) {
+            sub.unsubscribe();
+          }
+          const elementFinder = new ElementFinder(
+            destination.scopeChecker.namespace,
+            this.isolateModule,
+          );
+          const newElm = elementFinder.call()[0];
+          const newSub = fromEvent(
+            newElm,
+            type,
+            false,
+            false,
+            destination.passive,
+          ).subscribe({
+            next: event =>
+              this.onEvent(type, event, !!destination.passive, false),
+            error: () => {},
+            complete: () => {},
           });
+          insert(type, newElm, {sub: newSub, destination});
         } else {
-          this.removeDestination(id);
+          insert(type, elm, value);
         }
-      },
+      });
+      this.nonBubblingListeners = newMap;
     });
-    const destination: Destination = {id, selector, scopeChecker, subject};
-    this.destinations.push(destination);
-    return subject;
   }
 
-  /**
-   * Removes the destination that has the given id.
-   */
-  private removeDestination(id: number): void {
-    const i = indexOf(this.destinations, id);
-    i >= 0 && this.destinations.splice(i, 1); // tslint:disable-line:no-unused-expression
+  private buildPath(element: Element, namespace: Array<Scope>): Element[] {
+    const totalNamespace = getTotalIsolatedScope(namespace);
+    const root = this.isolateModule.getElement(totalNamespace);
+    let curr = element;
+    const result: Element[] = [element];
+    do {
+      curr = curr.parentNode as Element;
+      result.push(curr);
+    } while (curr && curr !== root);
+    return result;
   }
 
-  private capture(ev: Event) {
-    const n = this.destinations.length;
-    for (let i = 0; i < n; i++) {
-      const dest = this.destinations[i];
-      if (matchesSelector(ev.target as Element, dest.selector)) {
-        dest.subject._n(ev);
-      }
-    }
+  private onEvent(
+    type: string,
+    event: Event,
+    passive: boolean,
+    bubbles = true,
+  ): void {
```

Comment:
I'm worried about this function. It calls
- getTotalIsolatedScope (does slice + splice)
- buildPath: allocates an array and has a while loop
- slice and reverse
- bubble twice: which allocates a `Set` and has 3 nested loops: while-for-for

And it's on hot code path, so imagine a scroll event or a mousemove event on the DOM calling `onEvent` frequently.

Reply:
Should be fixed now

Reply:
Thanks! I'll take a look